java-frontend/src/main/java/org/sonar/java/ExecutionTimeReport.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java;
+
+import java.time.Clock;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.stream.Collectors;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+public class ExecutionTimeReport {
+  private static final Logger LOG = Loggers.get(ExecutionTimeReport.class);
+
+  private static final int MAX_REPORTED_FILES = 10;
+  private static final long MIN_REPORTED_ANALYSIS_TIME_MS = 1000;
+  private static final long MIN_TOTAL_ANALYSIS_TIME_TO_REPORT_MS = 20_000;
+
+  private static class ExecutionTime {
+    private final String file;
+    private final long analysisTime;
+
+    public ExecutionTime(String file, long analysisTime) {
+      this.file = file;
+      this.analysisTime = analysisTime;
+    }
+  }
+
+  private static final Comparator<ExecutionTime> ORDER_BY_ANALYSIS_TIME_DESCENDING_AND_FILE_ASCENDING = (a, b) -> {
+    int compare = -Long.compare(a.analysisTime, b.analysisTime);
+    if (compare == 0) {
+      compare = a.file.compareTo(b.file);
+    }
+    return compare;
+  };
+
+  private final LinkedList<ExecutionTime> recordedOrderedExecutionTime = new LinkedList<>();
+  private long minRecordedOrderedExecutionTime = MIN_REPORTED_ANALYSIS_TIME_MS;
+
+  private final Clock clock;
+  private final long analysisStartTimeMS;
+  private String currentFile;
+  private long currentFileStartTimeMS;
+
+  public ExecutionTimeReport(Clock clock) {
+    this.clock = clock;
+    analysisStartTimeMS = clock.millis();
+  }
+
+  public void start(String currentFile) {
+    this.currentFile = currentFile;
+    currentFileStartTimeMS = clock.millis();
+  }
+
+  public void end() {
+    long currentAnalysisTime = clock.millis() - currentFileStartTimeMS;
+    if (currentAnalysisTime >= minRecordedOrderedExecutionTime) {
+      recordedOrderedExecutionTime.add(new ExecutionTime(currentFile, currentAnalysisTime));
+      recordedOrderedExecutionTime.sort(ORDER_BY_ANALYSIS_TIME_DESCENDING_AND_FILE_ASCENDING);
+      if (recordedOrderedExecutionTime.size() > MAX_REPORTED_FILES) {
```

Comment:
To be honest I feel that we should filter out only at the end of the full analysis. We might be interested in seeing everything, maybe if set logs to DEBUG.

Reply:
My fear was to store a lot of data or manipulate too often the `recordedOrderedExecutionTime` list.
That's why there's this non-costly test `if (currentAnalysisTime >= minRecordedOrderedExecutionTime) {`
Once the list is full and the `minRecordedOrderedExecutionTime` strat to increase, it's super rare entering this test.
For example, on the ruling, the list contains only 3 or 4 elements and we never enter into this filter.

But I like the idea of logging more execution time in DEBUG mode. My proposal is to log each execution time, file by file, but probably more in TRACE mode than in DEBUG mode.